src/include/migraphx/rewrite_reshapes.hpp

```diff
@@ -0,0 +1,182 @@
+#ifndef MIGRAPHX_GUARD_MIGRAPHX_REWRITE_RESHAPES_HPP
+#define MIGRAPHX_GUARD_MIGRAPHX_REWRITE_RESHAPES_HPP
+
+#include <migraphx/config.hpp>
+#include <migraphx/matcher.hpp>
+#include <migraphx/make_op.hpp>
+#include <migraphx/pass_manager.hpp>
+#include <migraphx/common_dims.hpp>
+#include <migraphx/simplify_reshapes.hpp>
+#include <migraphx/eliminate_common_subexpression.hpp>
+#include <migraphx/dead_code_elimination.hpp>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+struct rewrite_reshapes_base
+{
+    template <class AxesMap>
+    static instruction_ref insert(module_pass_manager& mpm,
+                                  instruction_ref ins,
+                                  const std::vector<instruction_ref>& inputs,
+                                  const AxesMap&)
+    {
+        return mpm.get_module().insert_instruction(
+            ins, ins->get_operator(), inputs, ins->module_inputs());
+    }
+
+    template <class AxesMap>
+    static bool supports(instruction_ref, std::vector<std::size_t>&, const AxesMap&)
+    {
+        return true;
+    }
+
+    static std::vector<std::size_t> base_dims(instruction_ref ins)
+    {
+        return ins->get_shape().lens();
+    }
+};
+
+template <class T>
+struct rewrite_reshapes
+{
+    std::string name() const { return "rewrite_reshapes"; }
+    struct find_op_reshape_op
+    {
+        std::string op1;
+        std::string op2;
+
+        auto matcher() const
+        {
+            auto reshape =
+                match::name("reshape", "squeeze", "unsqueeze", "flatten")(match::used_once());
```

Comment:
add `reshape_lazy` too

Reply:
`reshape_lazy` doesn't work ?

Reply:
Maybe it requires another pass to add the copies back in and convert it to reshape after things are rewritten?

Reply:
There is no `reshape_lazy` until after lowering.